tsl/src/continuous_aggs/invalidation_plugin.c

```diff
@@ -0,0 +1,533 @@
+	 * when we can just ignore the tuple.
+	 */
+	Assert(!attr->attisdropped);
+	if (attr->attisdropped)
+		return;
+
+	bool isnull;
+	Datum datum = heap_getattr(tuple, relinfo->attnum, tupdesc, &isnull);
+
+	Assert(!isnull);
+	if (isnull)
+		return;
+
+	invalidation_cache_record_entry(data->invals_cache,
+									relinfo->hypertable_relid,
+									ts_time_value_to_internal(datum, attr->atttypid));
+}
+
+static void
+invalidations_record_insert(InvalidationsPluginData *data, TrackRelInfo *relinfo,
+							LogicalDecodingContext *ctx, Relation relation,
+							ReorderBufferChange *change)
+{
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.newtuple));
+}
+
+static void
+invalidations_record_update(InvalidationsPluginData *data, TrackRelInfo *relinfo,
+							LogicalDecodingContext *ctx, Relation relation,
+							ReorderBufferChange *change)
+{
+	/*
+	 * The field oldtuple is only set if any of the columns in the replica
+	 * identity are modified. Otherwise, oldtuple is set to NULL.
+	 *
+	 * This means that if oldtuple is NULL, it is sufficient to use newtuple
+	 * to decide what range that was updated.
+	 */
+	if (change->data.tp.oldtuple)
+		invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.oldtuple));
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.newtuple));
+}
+
+static void
+invalidations_record_delete(InvalidationsPluginData *data, TrackRelInfo *relinfo,
+							LogicalDecodingContext *ctx, Relation relation,
+							ReorderBufferChange *change)
+{
+	/*
+	 * We need a replica identity to deal with deletes. If the table does not
+	 * have a replica identity, we would have to invalidate the entire table,
+	 * so instead we throw an error if the delete does not have an oldtuple.
+	 */
+	if (!change->data.tp.oldtuple)
+		ereport(ERROR,
+				errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				errmsg("table is missing replica identity"),
+				errdetail("Table \"%s\" need to have a replica identity to handle DELETE.",
+						  RelationGetRelationName(relation)));
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.oldtuple));
+}
+
+static List *
+parse_output_parameters(List *options)
+{
+	ListCell *option;
+	List *relinfo_list = NIL;
+
+	/*
+	 * Options are pairs of relation names and the attribute to read.
+	 *
+	 * We do not check if the "key" is a hypertable since we do not want to
+	 * rely on the TimescaleDB metadata.
+	 */
+	foreach (option, options)
+	{
+		DefElem *elem = lfirst(option);
+		const char *name = elem->defname;
+#if PG16_LT
+		List *qname = stringToQualifiedNameList(name);
+#else
+		List *qname = stringToQualifiedNameList(name, NULL);
+#endif
+		TrackRelInfo *relinfo = palloc0(sizeof(TrackRelInfo));
+		RangeVar *hypertable_rv = makeRangeVarFromNameList(qname);
+		relinfo->hypertable_relid = RangeVarGetRelid(hypertable_rv, AccessShareLock, true);
```

Comment:
Why do we need a lock on the hypertable to read WAL entries? We're reading history anyway, so no lock should be needed.